### PR TITLE
Requested User Review

### PR DESCRIPTION
**Pull request by PhilippFischer0**
- MLP und Value Klassen sind jetzt im ./src/value.py Verzeichnis
- Funktion zum binären Einlesen, Downsamplen der Daten in ./src/__init__.py
- Der Rest in ./notebooks/binary_mnist.ipynb

### CODE REVIEW

File: src/value.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    # return the loss of a prediction in reference to the correct label
+    def cross_entropy_loss(y_pred: Value, y_gt) -> Value:
+        eps = 1e-15
+
+        if y_gt == 0:
+            return -((1 - y_pred + eps).log())
+        else:
+            return -((y_pred + eps).log())
+
+    def epoch_loss_and_accuracy(
```
Comment (Philipp36) [inline src/value.py:364]: Epochenloss und accuracy sollten unabhängig vom Modell bzw. dessen Komponenten sein.
Dein Modell soll generisch für diverse Probleme verwendbar sein, solange die Inputklasse (z.B. Bilder) und Outputklasse übereinstimmen.

File: .gitconfig [hunk 2]
```diff
@@ -1,3 +1,5 @@
-[filter "strip-notebook-output"]
-	clean = "jupyter nbconvert --ClearOutputPreprocessor.enabled=True --to=notebook --stdin --stdout --log-level=ERROR"
-	smudge = "jupyter nbconvert --ClearOutputPreprocessor.enabled=True --to=notebook --stdin --stdout --log-level=ERROR"
\ No newline at end of file
+[filter "nbstripout"]
+	clean = poetry run python -m nbstripout
+	smudge = cat
```
Comment (slavistan) [inline .gitconfig]: `cat` macht hier nichts. Kannst den smudge Filter ganz löschen.

File: src/value.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+        self, images: np.ndarray, labels: np.ndarray
+    ) -> tuple[float, float]:
+        loss = 0
+        correct_pred = 0
+        for image, label in zip(images, labels):
+            pred = self(image)
+
+            loss += Value.cross_entropy_loss(pred, label)
+
+            if np.fabs(pred.value - label.item()) < 0.5:
```
Comment (Philipp36) [inline src/value.py]: Ist zwar korrekt, allerdings würde ich hier noch eine Variable `y_pred` oder so einführen, wo du dann das von deinem Modell vorhergesagte Label zuweist. Das macht den Code mMn. lesbarer. Ich bin hier etwas drübergestolpert.

File: src/value.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+                correct_pred += 1.0
+
```
Comment (Philipp36) [inline src/value.py]: Hat das einen Grund, dass du hier explizit Fließkommaarithmetik verwendest?

File: src/value.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+    # logarithm
+    def log(self) -> Value:
+        result_value = np.log(self.value)
+        result = Value(result_value, (self,), name="log")
+
+        def _backward():
+            if self.value > 0:
+                self.grad += (1 / self.value) * result.grad
+            else:
+                self.grad += 0.0  # Gradient is zero for non-positive input
```
Comment (slavistan) [inline src/value.py]: Der Logarithmus ist undefiniert für x ≤ 0. Entsprechend ist seine Ableitung in diesem Intervall auch nicht definiert. Du kannst hier eine Warnung ausgeben oder einen Fehler werfen. Die Ableitung zu Null zu machen ist definitiv falsch 😉

Prüfe die Bedingung direkt am Anfang der Funktion. Damit brauchst du auch die Prüfung in `_backward` nicht.

File: src/value.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+        def _backward():
+            self.grad += -result.grad
+
+        result._backward = _backward
+        return result
+
+    # activation functions
+    def relu(self) -> Value:
+        result_value = self.value if self.value > 0 else 0.0
+        result = Value(result_value, (self,), name="ReLU")
```
Comment (slavistan) [inline src/value.py]: "relu"; wähle eine einheitliche Schreibweise

File: src/value.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+        def _backward():
+            self.grad += result.value * result.grad
+
+        result._backward = _backward
+        return result
+
+    # logarithm
+    def log(self) -> Value:
+        result_value = np.log(self.value)
```
Comment (slavistan) [inline src/value.py]: unnötige Zwischenzuweisung; Erzeuge deinen Value direkt.

File: src/value.py [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    def load_params(self, file_path: str) -> None:
+        with open(file_path, "rb") as file:
+            param = pickle.load(file)
+
+        for (_, param_value), p in zip(param, self.parameters()):
+            p.value = param_value
+
+    @staticmethod
+    # return the loss of a prediction in reference to the correct label
+    def cross_entropy_loss(y_pred: Value, y_gt) -> Value:
```
Comment (Philipp36) [inline src/value.py:350]: Hier gerne noch die Clipping Logik für den Fall `y_pred + eps > 1` ergänzen.

File: src/value.py [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+
+    # return all parameters for every single Neuron inside of a Layer
+    def parameters(self) -> list:
+        params = [p for n in self.neurons for p in n.parameters()]
+        return params
+
+
+class MLP:
+    def __init__(self, num_inputs: int, num_hidden: list[int], num_out: int) -> None:
+        size = [num_inputs] + num_hidden
```
Comment (Philipp36) [inline src/value.py]: Gerne in `sizes`, `layer_sizes` oder ähnliches umbenennen. Der Singular suggeriert, dass sich nur ein Wert hinter der Variable verbirgt.

File: src/value.py [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+                else:
+                    dot.edge(ancestor_name, unique_node_name)
+
+            return unique_node_name
+
+        add_nodes(dot, self)
+        display(dot)
+
+
+np.random.seed(0xDEADBEEF)
```
Comment (slavistan) [inline src/value.py]: Seeds werden im Applikationscode (Script oder Notebook) gesetzt, nicht versteckt in Bibliotheken; kann weg.

File: src/value.py [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+        self.bias = Value(0.0, name="bias")
+
+    def __call__(self, x: np.ndarray) -> Value:
+        # f(x) = activation (bias + sum(weights * values))
+        if isinstance(x, np.ndarray):
+            x = x.flatten()
+        res = sum(w_i * x_i for w_i, x_i in zip(self.weights, x)) + self.bias
+        return res
+
+    # return the parameters of each Neuron
```
Comment (slavistan) [inline src/value.py]: Wenn du Funktionen oder Klassen mit Kommentaren versehen möchtest, nimm Docstrings, ist ein üblicher Brauch bei der Pythonprogrammierung.

```python
def my_func():
    """
    My Func does funky things.
    """

    print("hello world")
```

File: src/value.py [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+
+    def save_params(self, file_path: str) -> None:
+        params = [(p.name, p.value) for p in self.parameters()]
+
+        with open(file_path, "wb") as file:
+            pickle.dump(params, file)
+
+    def load_params(self, file_path: str) -> None:
+        with open(file_path, "rb") as file:
+            param = pickle.load(file)
```
Comment (Philipp36) [inline src/value.py]: Auch hier Singular -> Plural

File: src/value.py [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+        def _backward():
+            self.grad += sigmoid_value * (1 - sigmoid_value) * result.grad
+
+        result._backward = _backward
+        return result
+
+    # comparison operators
+    # less than <
+    def __lt__(self, other: Value) -> bool:
```
Comment (slavistan) [inline src/value.py]: Typannotationen sind unvollständig: `other` kann im Moment nur ein `Value` sein, du erwartest aber der Definition deiner Funktion nach auch andere Typen, vermutlich `float` bzw. etwas "Zahlenartiges".

File: src/value.py [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+
+class Neuron:
+    def __init__(self, num_inputs: int) -> None:
+        self.weights = [
+            Value(np.random.randn(), name="weight") for _ in range(num_inputs)
+        ]
+        self.bias = Value(0.0, name="bias")
+
+    def __call__(self, x: np.ndarray) -> Value:
+        # f(x) = activation (bias + sum(weights * values))
```
Comment (slavistan) [inline src/value.py]: wird das gebraucht?

File: .gitattributes
Comment (Philipp36) [inline .gitattributes:1]: Wird die Datei noch benötigt?

File: pyproject.toml
Comment (Philipp36) [inline pyproject.toml:1]: Splitte deine Dependencies bitte noch in Gruppen auf, dann sieht es aufgeräumter aus. Wie wäre es mit einer `dev` oder `development` Gruppe bspw.?
